luigi/contrib/pai.py

```diff
@@ -0,0 +1,300 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""
+MicroSoft OpenPAI Job wrapper for Luigi.
+
+  "OpenPAI is an open source platform that provides complete AI model training and resource management capabilities,
+  it is easy to extend and supports on-premise, cloud and hybrid environments in various scale."
+
+For more information about OpenPAI : https://github.com/Microsoft/pai/, this task is tested against OpenPAI 0.7.1
+
+Requires:
+
+- requests: ``pip install requests``
+
+Written and maintained by Liu, Dongqing (@liudongqing).
+"""
+import time
+import logging
+import luigi
+import abc
+
+try:
+    from urlparse import urljoin
+except ImportError:
+    from urllib.parse import urljoin
+
+import json
+
+logger = logging.getLogger('luigi-interface')
+
+try:
+    import requests as rs
+
+except ImportError:
+    logger.warning('requests is not installed. PaiTask requires requests.')
+
+
+def slot_to_dict(o):
+    o_dict = {}
+    for key in o.__slots__:
+        if not key.startswith('__'):
+            value = getattr(o, key, None)
+            if value is not None:
+                o_dict[key] = value
+    return o_dict
+
+
+class PaiJob(object):
+    """
+    The Open PAI job definition.
+    Refer to here https://github.com/Microsoft/pai/blob/master/docs/job_tutorial.md
+    ::
+
+        {
+          "jobName":   String,
+          "image":     String,
+          "authFile":  String,
+          "dataDir":   String,
+          "outputDir": String,
+          "codeDir":   String,
+          "virtualCluster": String,
+          "taskRoles": [
+            {
+              "name":       String,
+              "taskNumber": Integer,
+              "cpuNumber":  Integer,
+              "memoryMB":   Integer,
+              "shmMB":      Integer,
+              "gpuNumber":  Integer,
+              "portList": [
+                {
+                  "label": String,
+                  "beginAt": Integer,
+                  "portNumber": Integer
+                }
+              ],
+              "command":    String,
+              "minFailedTaskCount": Integer,
+              "minSucceededTaskCount": Integer
+            }
+          ],
+          "gpuType": String,
+          "retryCount": Integer
+        }
+
+    """
+    __slots__ = (
+        'jobName', 'image', 'authFile', 'dataDir', 'outputDir', 'codeDir', 'virtualCluster',
+        'taskRoles', 'gpuType', 'retryCount'
+    )
+
+    def __init__(self, jobName, image, tasks):
+        self.jobName = jobName
```

Comment:
Can this `__init__` get a docstring for input expectation descriptions?

Reply:
Sure